pkg/cmd/codespace/ssh.go

```diff
@@ -147,6 +155,129 @@ func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err e
 	}
 }
 
+func (a *App) ListOpensshConfig(ctx context.Context, opts configOptions) error {
+	// Ensure all child tasks (e.g. port forwarding) terminate before return.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var err error
+	var codespaces []*api.Codespace
+	if opts.codespace == "" {
+		a.StartProgressIndicatorWithLabel("Fetching codespaces")
+		codespaces, err = a.apiClient.ListCodespaces(ctx, -1)
+		a.StopProgressIndicator()
+	} else {
+		var codespace *api.Codespace
+		codespace, err = getOrChooseCodespace(ctx, a.apiClient, opts.codespace)
+		codespaces = []*api.Codespace{codespace}
+	}
+	if err != nil {
+		return fmt.Errorf("error getting codespace info: %w", err)
+	}
+
+	t, err := template.New("ssh_config").Parse(`Host cs.{{.Name}}.{{.EscapedRef}}
+	User {{.SshUser}}
+	ProxyCommand {{.GHExec}} cs ssh -c {{.Name}} --stdio
+	UserKnownHostsFile=/dev/null
+	StrictHostKeyChecking no
+	LogLevel quiet
+	ControlMaster auto
+
+`)
+	if err != nil {
+		return fmt.Errorf("error formatting template: %w", err)
+	}
+
+	ghexec, err := os.Executable()
+	if err != nil {
+		return err
+	}
+
+	sshUsers := map[string]string{}
+
+	for _, cs := range codespaces {
+
+		if cs.State != "Available" {
+			fmt.Fprintf(os.Stderr, "skipping unavailable codespace %s: %s\n", cs.Name, cs.State)
+			continue
+		}
+
+		var sshUser string
+		var ok bool
+
+		if sshUser, ok = sshUsers[cs.Repository.FullName]; !ok {
+			session, err := openSshSession(ctx, a, cs.Name, nil)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error connecting to codespace: %v\n", err)
+				continue
+			}
+			defer session.Close()
+
+			a.StartProgressIndicatorWithLabel(fmt.Sprintf("Fetching SSH Details for %s", cs.Name))
+			_, sshUser, err = session.StartSSHServer(ctx)
+			a.StopProgressIndicator()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error getting ssh server details: %v", err)
+				continue
+			}
+			sshUsers[cs.Repository.FullName] = sshUser
+		}
+
+		conf := codespaceSshConfig{
+			Name:       cs.Name,
+			EscapedRef: strings.ReplaceAll(cs.GitStatus.Ref, "/", "-"),
+			SshUser:    sshUser,
+			GHExec:     ghexec,
+		}
+		if err := t.Execute(a.io.Out, conf); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type codespaceSshConfig struct {
+	Name       string
+	EscapedRef string
+	SshUser    string
+	GHExec     string
+}
+
+func openSshSession(ctx context.Context, a *App, csName string, liveshareLogger *log.Logger) (*liveshare.Session, error) {
+	codespace, err := getOrChooseCodespace(ctx, a.apiClient, csName)
```

Comment:
This function call duplicates expensive work already done; move it into the caller and have it pass in the codespace struct (not name).

Reply:
Unfortunately, different codespaces can have different remote ssh usernames so we need to connect to them separately. I added a comment explaining this. I at least build a map of repository -> username so that we only need to do this once per repository.

Reply:
Perhaps more like [this](https://github.com/cli/cli/pull/4915/commits/a9d02a746baa3b3cfa17d7c9cd1f80e030cb4f65)?